### PR TITLE
Improved pooling logic for EMR runner

### PR DESCRIPTION
**Pull request by aaronabf**
This achieves two features

1. We add an option `bypass_pool_wait` that if set will ignore the `pool_wait_minutes` time if there are no clusters with a valid setup. See ticket [DATA-258](https://jira.team.affirm.com/browse/DATA-258)
2. Mirrors logic I implement in the YARN runner to reduce number of EMR API calls. This uses my previous logical split of cluster setup and cluster state, and only rechecks the state of cluster for each pooling check.
3. Lastly this fixes a dumb bug in mrjob where they set accidentally set `num_steps` twice and totally mess up the logic.

### CODE REVIEW

File: mrjob/emr.py [hunk 1]
```diff
@@ -2549,7 +2550,8 @@ def _check_cluster_state(self, emr_client, cluster, num_steps):
 
         return len(steps)
 
-    def _usable_clusters(self, exclude=set(), num_steps=1):
+    def _usable_clusters(self, valid_clusters, invalid_clusters,
+                         locked_clusters, num_steps):
```
Comment (ejsundstr) [inline mrjob/emr.py:2555]: Note a huge fan of pass by reference in python.

File: mrjob/emr.py [hunk 2]
```diff
@@ -2603,45 +2619,52 @@ def _find_cluster(self, num_steps=1):
...
         end_time = now + timedelta(minutes=max_wait_time)
-        time_sleep = timedelta(seconds=_POOLING_SLEEP_INTERVAL)
 
         log.info('Attempting to find an available cluster...')
         while now <= end_time:
             cluster_info_list = self._usable_clusters(
-                exclude=exclude,
-                num_steps=num_steps)
+                valid_clusters, invalid_clusters,
+                locked_clusters, num_steps)
```
Comment (ejsundstr) [inline mrjob/emr.py:2637]: I don't know if they use asserts anywhere in this code but I would add 
`assert len(valid_clusters & invalid_clusters) == 0`

or log or whatever, just check that your assumptions are met.

**Review by ejsundstr — APPROVED** [review]
Left some comments but they aren't major.

**Comment by aaronabf** [comment]
Thanks for the feedback!

File: mrjob/emr.py [hunk 3]
```diff
@@ -2558,9 +2561,16 @@ def _usable_clusters(self, exclude=set(), num_steps=1):
         - total compute units for master node
         - time left to an even instance hour
 
-        The most desirable clusters come *last* in the list.
+        :param valid_clusters: A map of cluster id to cluster info with a valid
```
Comment (gregsterin) [inline mrjob/emr.py:2564]: Worth noting that this function side-effects valid_clusters, invalid_clusters (it can add new items to these lists).

**Pull request merged**